### PR TITLE
fix: url validation

### PR DESCRIPTION
**Pull request by saikumarrs**
## PR Description

Fixed the URL validation utility to correctly flag invalid/incomplete URLs with the help of regex validation.
I've made sure that the same utility is consistently used across all the user touchpoints in the SDK to validate the URL inputs.
Moreover, the utility is now moved to the common package so that it can be used across the projects.

Other minor updates:
- All the dependencies are upgraded to the latest versions.
- The dev and prod dependencies are re-defined in the root and individual projects` package.json files.
  - This will ensure dependencies are properly managed in the monorepo and help us in correctly identify the affected packages. Nx also recommends this.
  - With this, we don't have to define prod dependencies as dev anymore.
- All the workflows are cleaned up and formatted.
  - The job constraints for branch names are fixed.
  - Removed job constraints for certain workflows where it is dependent on a previous job that has identical constraints.
  - For the NPM publish action, we're now running to script to clean up unwanted fields from the package.json so that they don't end up in NPM. This is mainly done for the `dependencies` and `devDependencies` fields.
  - Slack notifications are tailored for different deployment environments.
- Replace `api.rudderlabs.com` with `api.rudderstack.com` in all the example applications.
- Fixed sanity suite to correctly compare the test data.

For more insights, see "Summary by CodeRabbit" section below.

## Linear task (optional)

https://linear.app/rudderstack/issue/SDK-1894/validate-the-data-plane-url-correctly-js-sdk

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [ ] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [ ] All sanity suite test cases pass locally

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved URL validation across various components.
  - Fixed issues with deployment workflows to ensure smoother CI/CD processes.

- **Refactor**
  - Updated deployment workflows to streamline conditions and environment variable handling.
  - Replaced `isValidUrl` with `isValidURL` for better consistency and reliability.

- **Chores**
  - Adjusted size limits for specific modules to accommodate recent changes.
  - Updated Slack notification messages for new releases.

- **Tests**
  - Enhanced test cases for URL validation and capabilities management.
  - Removed redundant tests for deprecated URL validation functions.

### CODE REVIEW

File: .github/workflows/deploy-sanity-suite.yml [hunk 1]
```diff
@@ -132,19 +132,19 @@ jobs:
 
       - name: Sync files to S3
         run: |
-          aws s3 cp packages/sanity-suite/dist/ s3://${{ env.AWS_S3_BUCKET_NAME }}/sanity-suite/${{ env.SUITE_CDN_PATH }} --recursive --cache-control max-age=3600
+          aws s3 cp packages/sanity-suite/dist/ s3://${{ env.AWS_S3_BUCKET_NAME }}/sanity-suite${{ env.SUITE_CDN_PATH }} --recursive --cache-control max-age=3600
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply .github/workflows/deploy-sanity-suite.yml:133]: No action required here.

File: packages/analytics-js/src/components/configManager/ConfigManager.ts [hunk 2]
```diff
@@ -103,7 +104,7 @@ class ConfigManager implements IConfigManager {
         const dataServiceUrl = getDataServiceUrl(
           dataServiceEndpoint ?? DEFAULT_DATA_SERVICE_ENDPOINT,
         );
-        if (isValidUrl(dataServiceUrl)) {
+        if (isValidURL(dataServiceUrl)) {
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/src/components/configManager/ConfigManager.ts:107]: This is not in the scope of the current task.

File: packages/analytics-js/__tests__/components/capabilitiesManager/CapabilitiesManager.test.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+      expect(consoleWarn).not.toHaveBeenCalled();
+    });
+
+    it('should not load polyfills if default polyfill URL is invalid', () => {
+      state.loadOptions.value.polyfillURL = 'invalid-url';
+      state.lifecycle.writeKey.value = 'sample-write-key';
+      state.loadOptions.value.polyfillIfRequired = true;
+
+      POLYFILL_URL = 'invalid-url';
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/capabilitiesManager/CapabilitiesManager.test.ts:129]: This is needed for the test case.

File: packages/analytics-js/__tests__/components/configManager/cdnPaths.test.ts [hunk 4]
```diff
@@ -15,52 +18,87 @@ jest.mock('../../../src/components/configManager/util/commonUtil.ts', () => {
...
+    it('should throw error if invalid custom url is provided', () => {
+      const pluginsCDNPath = () => getPluginsCDNPath('htp:/some.broken.url');
+      expect(pluginsCDNPath).toThrow(
+        'Failed to load the SDK as the CDN base URL for plugins is not valid.',
+      );
+    });
+
+    it('should return default plugins CDN URL if no custom URL is provided and no script src exists', () => {
+      const pluginsCDNPath = getPluginsCDNPath(undefined);
+      expect(pluginsCDNPath).toBe(`https://cdn.rudderlabs.com/v3/modern/plugins`);
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/configManager/cdnPaths.test.ts]: Done.

File: packages/analytics-js/__tests__/components/capabilitiesManager/CapabilitiesManager.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/capabilitiesManager/CapabilitiesManager.test.ts:129]: It doesn't work.

[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/capabilitiesManager/CapabilitiesManager.test.ts:129]: I don't think you can mock an import within a test case.

File: packages/sanity-suite/src/testBook/TestBook.js [hunk 5]
```diff
@@ -299,10 +299,10 @@ class TestBook {
       const observer = new MutationObserver(mutationList => {
         const resultDataElement = mutationList[0].addedNodes[0].parentNode;
         const resultData = resultDataElement.textContent.trim();
-        // Get the last but one child from resultRowElement
+        // Get the last but second child from resultRowElement
         const expectedResult =
           resultRowElement.childNodes[
-            resultRowElement.childNodes.length - 2
+            resultRowElement.childNodes.length - 4
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/sanity-suite/src/testBook/TestBook.js]: Done.

File: scripts/make-package-json-publish-ready.sh [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+# Define the directory containing your packages
+PACKAGES_DIR="$BASE_DIR/packages"
+
+# Iterate over each package directory
+for package in "$PACKAGES_DIR"/*; do
+  if [ -d "$package" ]; then
+    PACKAGE_JSON="$package/package.json"
+    if [ -f "$PACKAGE_JSON" ]; then
+      echo "Cleaning $PACKAGE_JSON..."
+      jq 'del(.dependencies, .devDependencies, .peerDependencies, .optionalDependencies, .overrides, .scripts, .browserslist)' "$PACKAGE_JSON" > "$package/package_cleaned.json" && mv "$package/package_cleaned.json" "$PACKAGE_JSON"
```
Comment (MoumitaM) [inline scripts/make-package-json-publish-ready.sh:14]: Can you explain this command? what is it doing?

Comment (saikumarrs) [reply scripts/make-package-json-publish-ready.sh:14]: We're deleting certain fields in the JSON file.

Comment (MoumitaM) [reply scripts/make-package-json-publish-ready.sh:14]: Is it deleting all the dependencies in the package.json?

Comment (saikumarrs) [reply scripts/make-package-json-publish-ready.sh:14]: Yes. We don't need anything for the published package.

**Review by MoumitaM — DISMISSED** [review]

**Pull request merged**